File: Jin_Vision/plateSub.py
import datetime
import imutils
import time
import cv2
import numpy as np
from backgroundAveraging import BackGroundAveraging
from shapeDetector import ShapeDetector
from colourDetector import ColourDetector
from colourContain import ColourContain

class PlateSub:

	def __init__(self,DEBUG,backgroundFrame):
		self.backgroundFrame = backgroundFrame.copy()
		self.DEBUG = DEBUG


	def image_processing(self,currentFrame):
		np.set_printoptions(threshold=np.nan)
		currentFrame_copy = currentFrame.copy()
		# if currentFrame_copy.shape[0] < self.backgroundFrame.shape[0]:
		# 	height = currentFrame_copy.shape[0]

		# 	if currentFrame_copy.shape[1] < self.backgroundFrame.shape[1]:
		# 		width = currentFrame_copy.shape[1]
		# 	elif currentFrame_copy.shape[1] >= self.backgroundFrame.shape[1]:
		# 		width = self.backgroundFrame.shape[1]

		# elif currentFrame_copy.shape[0] >= self.backgroundFrame.shape[0]:
		# 	height = self.backgroundFrame.shape[0]

		# 	if currentFrame_copy.shape[1] < self.backgroundFrame.shape[1]:
		# 		width = currentFrame_copy.shape[1]
		# 	elif currentFrame_copy.shape[1] >= self.backgroundFrame.shape[1]:
		# 		width = self.backgroundFrame.shape[1]

		width = min(currentFrame_copy.shape[0], self.backgroundFrame.shape[0])
		height = min(currentFrame_copy.shape[1], self.backgroundFrame.shape[1])

		# currentFrame_copy = imutils.resize(currentFrame_copy, height=height,width=width)
		# self.backgroundFrame = imutils.resize(self.backgroundFrame, height=height,width=width)
		currentFrame_copy = np.resize(currentFrame_copy, (height,width,3))
		self.backgroundFrame = np.resize(self.backgroundFrame, (height,width,3))

		currentGray = cv2.cvtColor(currentFrame_copy,cv2.COLOR_BGR2GRAY)
		currentGray = cv2.GaussianBlur(currentGray,(1,1),0)

		backGroundGray = cv2.cvtColor(self.backgroundFrame,cv2.COLOR_BGR2GRAY)
		backGroundGray = cv2.GaussianBlur(backGroundGray,(1,1),0)

		if(self.DEBUG):
			cv2.imshow("CurrentGaussianBlur",currentGray)
			cv2.imshow("BackGroundGaussianBlur",backGroundGray)
			cv2.waitKey(0)

		print currentGray.shape
		print backGroundGray.shape
		cv2.accumulateWeighted(currentGray,backGroundGray.astype("float"),0.5)
		frameDelta = cv2.absdiff(cv2.convertScaleAbs(backGroundGray),currentGray)
		#frameDelta = cv2.absdiff(backGroundGray,currentGray)

		if(self.DEBUG):
			cv2.imshow("frameDelta",frameDelta)
			cv2.waitKey(0)

		#set the delta is less than 50, we discard the pixel and set it to black. 
		#If the delta is greater than 50, we will set it to white
		if(self.DEBUG):
			print frameDelta
		thresh = cv2.threshold(frameDelta,50,255,cv2.THRESH_BINARY)[1]
		if(self.DEBUG):
			cv2.imshow("thresh1",thresh)
			cv2.waitKey(0)

		thresh = cv2.dilate(thresh,None,iterations=1)
		if(self.DEBUG):
			cv2.imshow("thresh2",thresh)
			cv2.waitKey(0)

		filter_image = np.zeros((currentFrame_copy.shape),dtype="uint8")
		for i in range (0,thresh.shape[0]):
			for j in range (0,thresh.shape[1]):
				if thresh[i,j] == 255:
					filter_image[i,j] = currentFrame_copy[i,j].copy()

		if(self.DEBUG):
			cv2.imshow("filter_image",filter_image)
			cv2.waitKey(0)

		cp = ColourContain(self.DEBUG)
		color_plate = cp.colour_contain(filter_image)

		if(self.DEBUG):
			print "The plate contains "
			print color_plate

		return color_plate




		# [_,cnts,_] = cv2.findContours(thresh.copy(),cv2.RETR_EXTERNAL,cv2.CHAIN_APPROX_SIMPLE)

		# position = []
		# sd = ShapeDetector()
		# cl = ColourDetector()
		# for c in cnts:
		# 	#set minimum number of pixels that considers object
		# 	if cv2.contourArea(c) < 100:
		# 		if(self.DEBUG):
		# 			print"The area of contour is %d"%(cv2.contourArea(c))
		# 		continue

		# 	M = cv2.moments(c)
		# 	cX = int((M["m10"] / M["m00"]))
		# 	cY = int((M["m01"] / M["m00"]))
		# 	shape = sd.detect(c)
		# 	color = cl.colour_detect(currentLab,c)
		# 	c = c.astype("int")
		# 	cv2.drawContours(currentFrame_copy,[c],-1,(0,255,0),2)
		# 	cv2.putText(currentFrame_copy, color, (cX, cY),cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 2)
		# 	cv2.imshow("Object Detection",currentFrame_copy)
		# 	cv2.waitKey(0)
		# return currentFrame_copy,position

File: fred/src/backgroundSub/PlateSub.java
package backgroundSub;

import org.opencv.core.*;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by jinhong on 09/03/2017.
 */
public class PlateSub implements BackGroundSub {

    private Boolean DEBUG;
    private Mat originalImage = new Mat();
    private ColourContain cp = new ColourContain(true);
    public Size size = new Size(1,1);
    public Imshow imshow1 = new Imshow("CurrentGaussianBlur");
    public Imshow imshow2 = new Imshow("BackGroundGaussianBlur");
    public Imshow imshow3 = new Imshow("frameDelta");
    public Imshow imshow4 = new Imshow("thresh1");
    public Imshow imshow5 = new Imshow("thresh2");
    public Imshow imshow6 = new Imshow("filter image");

    public PlateSub(Boolean DEBUG,Mat originalImage)
    {
        this.DEBUG = DEBUG;
        Imgproc.resize(originalImage,this.originalImage,new Size(30,30));
    }

    @Override
    public List<String> image_processing(Mat frame) {
        Mat currentGray = new Mat(),backgroundGray = new Mat(),frameDelta = new Mat(),thresh = new Mat();
        List<String> colour_plate = new ArrayList<String>();

        Mat filter_image = new Mat(30, 30, CvType.CV_8UC3, new Scalar(0,0,0));

        Mat resized_frame = new Mat();
        Imgproc.resize(frame,resized_frame,new Size(30,30));


        Imgproc.cvtColor(resized_frame, currentGray, Imgproc.COLOR_BGR2GRAY);
        Imgproc.GaussianBlur(currentGray, currentGray, size, 0);

        Imgproc.cvtColor(this.originalImage, backgroundGray, Imgproc.COLOR_BGR2GRAY);
        Imgproc.GaussianBlur(backgroundGray, backgroundGray, size, 0);

        if (this.DEBUG) {
            imshow1.showImage(currentGray);
            imshow2.showImage(backgroundGray);
        }

//        Imgproc.accumulateWeighted(currentGray, backgroundGray, 0.5);
        Core.convertScaleAbs(backgroundGray, backgroundGray);
        Core.absdiff(backgroundGray, currentGray, frameDelta);

        if (this.DEBUG) {
            imshow3.showImage(frameDelta);
        }

        Imgproc.threshold(frameDelta, thresh, 50, 255, Imgproc.THRESH_BINARY);

        if (this.DEBUG) {
            imshow4.showImage(thresh);
        }

        Imgproc.dilate(thresh, thresh, new Mat(), new Point(-1, -1), 1);

        if (this.DEBUG) {
            imshow5.showImage(thresh);
        }

        for(int i = 0; i < thresh.cols(); i++)
        {
            for(int j = 0; j < thresh.rows(); j++)
            {
                if(thresh.get(j,i)[0] == 255)
                {
                    filter_image.put(j,i,resized_frame.get(j,i));
                }
            }
        }
        if(this.DEBUG)
        {
            imshow6.showImage(filter_image);
        }
        colour_plate = cp.colour_contain(filter_image);

        return colour_plate;
    }
}
